src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,328 @@
+               program_invocation_short_name);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int get_known_solutions(sd_device *device, char ***ret_solutions) {
+        _cleanup_free_ char **s = NULL;
+        const char *value;
+        int r;
+
+        assert(ret_solutions);
+
+        r = sd_device_get_property_value(device, "IOCOST_SOLUTIONS", &value);
+        if (r < 0)
+                return r;
+
+        s = strv_split(value, " ");
+        if (!s)
+                return -ENOMEM;
+
+        *ret_solutions = TAKE_PTR(s);
+
+        return 0;
+}
+
+static int choose_solution(char **solutions, const char **ret_name) {
+        assert(ret_name);
+
+        if (strv_isempty(solutions))
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EINVAL), "IOCOST_SOLUTIONS exists in hwdb but is empty.");
+
+        if (arg_target_solution && strv_find(solutions, arg_target_solution)) {
+                *ret_name = arg_target_solution;
+                log_debug("Selected solution based on target solution: %s", *ret_name);
+        } else {
+                *ret_name = solutions[0];
+                log_debug("Selected first available solution: %s", *ret_name);
+        }
+
+        return 0;
+}
+
+static int query_named_solution(
+                sd_device *device,
+                const char *name,
+                const char **ret_model,
+                const char **ret_qos) {
+
+        _cleanup_strv_free_ char **solutions = NULL;
+        _cleanup_free_ char *upper_name = NULL, *qos_key = NULL, *model_key = NULL;
+        const char *qos = NULL, *model = NULL;
+        int r;
+
+        assert(ret_qos);
+        assert(ret_model);
+
+        /* If NULL is passed we query the default solution, which is the first one listed
+         * in the SOLUTIONS key or the one specified by the TargetSolution setting.
+         */
+        if (!name) {
+                r = get_known_solutions(device, &solutions);
+                if (r == -ENOENT) {
+                       log_debug_errno(r, "No entry found for device, skipping iocost logic.");
+                        return 0;
```

Comment:
we have this rule in systemd: when a function fails it should not touch any of its return parameters, and if it succeeds it should initialize *all* of them. failure is a return code of < 0, success anything else. here you do "return 0", hence this is a success case, hence please add `*ret_qos = *ret_model = NULL;` here. simply as a matter of sticking to our usual coding style and making this less surprising

Reply:
I know you added setting of the the output params here following the suggestion, but it's not useful ;(
When `query_named_solution()` does not find something, the caller has nothing to do. You may just return -ENOENT here.